metrics/constants.go

```diff
@@ -0,0 +1,39 @@
+package metrics
+
+import (
+	"gx/ipfs/QmNVpHFt7QmabuVQyguf8AbkLDZoFh7ifBYztqijYT1Sd2/go.opencensus.io/stats"
+	"gx/ipfs/QmNVpHFt7QmabuVQyguf8AbkLDZoFh7ifBYztqijYT1Sd2/go.opencensus.io/stats/view"
+	"gx/ipfs/QmNVpHFt7QmabuVQyguf8AbkLDZoFh7ifBYztqijYT1Sd2/go.opencensus.io/tag"
+)
+
+// KeyMethod is used to identified the method stats are collected for.
+var KeyMethod, _ = tag.NewKey("method")
+
+// Opencensus stats
+var (
+	// ProcessBlock duration in milliseconds
+	MProcessBlockMs = stats.Float64("consensus/process_block", "The duration in milliseconds of ProcessBlock", stats.UnitMilliseconds)
+
+	processBlockView = &view.View{
+		Name:        "consensus/process_block",
+		Measure:     MProcessBlockMs,
+		Description: "The distribution of the durations",
+
+		// Latency in buckets:
+		// [>=0ms, >=25ms, >=50ms, >=75ms, >=100ms, >=200ms, >=400ms, >=600ms, >=800ms, >=1s, >=2s, >=4s, >=8s]
+		Aggregation: view.Distribution(25, 50, 75, 100, 200, 400, 600, 800, 1000, 2000, 4000, 8000),
+		TagKeys:     []tag.Key{KeyMethod}}
+
+	// ApplyMessage duration in milliseconds
+	MApplyMessageMs = stats.Float64("consensus/apply_message", "The duration in milliseconds of ApplyMessage", stats.UnitMilliseconds)
+
+	applyMessageView = &view.View{
+		Name:        "consensus/apply_message",
+		Measure:     MApplyMessageMs,
+		Description: "The distribution of the durations",
+
+		// Latency in buckets:
+		// [>=0ms, >=25ms, >=50ms, >=75ms, >=100ms, >=200ms, >=400ms, >=600ms, >=800ms, >=1s, >=2s, >=4s, >=8s]
+		Aggregation: view.Distribution(25, 50, 75, 100, 200, 400, 600, 800, 1000, 2000, 4000, 8000),
+		TagKeys:     []tag.Key{KeyMethod}}
+)
```

Comment:
Blocking: this seems to be doing multiple things, it's not sufficiently documented what they are, and I think too centralised/static.

I infer that the `MProcessBlockMs` and `MApplyMessageMs` are definitions of the things that actually track metrics, "measures", and are referenced from `processor.go`. We should not have an explicit global list of all of these measures that's imported by everything in the code in this way. Better would be for the metrics to be defined by the modules exporting them. Even better would be for them to be less static so that they can be faked, tested and so the other stuff here doesn't need to import the world either (NB actually doing any of that faking and testing is way outside scope here).

I infer that the "View" stuff is about how values are exported to Prometheus. Is that right? It's not so bad for that to be centralised/static, but I do think we could do a lot better than all this boilerplate. The name and description belong with the metric definition in the implementing module. I would suggest for now that we can put the aggregation stuff there for now too and defer more complex aggregations to later.

So my summary comes down to moving almost all of this into processor.go. I recommend a helper function to remove the boilerplate and repetition of Measure/View construction, so a caller can do something like:

```
// In processor.go and other modules
// Under the hood this registers a view for exporting with prometheus.
pbLatency := metrics.Registry().NewTimer("consensus/process_block", "The duration in milliseconds of ProcessBlock", stats.UnitMilliseconds) // default aggregations and tags
```

An ideal outcome is that almost none of the go-filecoin code imports `opencensus` directly, but I can't quite tell whether that's reasonable yet.

Reply:
I will prototype something like this and push it up for discussion